### ISSUE TITLE
Daemon mode is slow to write trailing newline in response

### ISSUE THREAD

**Issue by FilipVozar** [VowpalWabbit/vowpal_wabbit#2146]
#### Describe the bug
When reading TCP responses from the vw daemon in cb_explore_adf, it looks like it sends the trailing newline in a separate network write, 40+ms after the result is sent. When sending the requests on the connection one by one (request, response, request, response, and so on), this significantly increases the latency - the result itself is received usually in <0.5ms, but then we have to wait 40+ms for the trailing newline to send a new request. This doesn't happen for the first response, but it happens for all the following responses.

This is an issue for us when serving low-latency production traffic. Currently we work around it by opening a new connection for each request, and closing it immediately after response is received.

#### Steps To Reproduce
Running vw daemon:
`vw  --cb_explore_adf --daemon --port=23456 --foreground  -q CV -t`

Minimum example code that demonstrates this behavior (in Go - I've tried to write similar network code in Python and got the same result):
```
package main

import (
	"bufio"
	"fmt"
	"net"
	"time"
)

func main() {
	conn, _ := net.Dial("tcp", "127.0.0.1:23456")
	r := bufio.NewReader(conn)

	for i := 0; i < 5; i++ {
		fmt.Println("-----")
		_, _ = conn.Write([]byte("shared |Context c:1.0\n|Action a:1.0\n\n"))

		for newlines := 0; newlines < 2; {
			t1 := time.Now()
			b, _ := r.ReadByte()
			fmt.Println("recv", b, time.Since(t1))
			if b == '\n' {
				newlines++
			}
		}
	}
}
```

#### Expected behavior
I would expect that the VW daemon would send the trailing newlines in a single network write call, like it does for the first response.

#### Observed Behavior
Output of the Go code above:
```
-----
recv 48 445.007µs
recv 58 77ns
recv 49 57ns
recv 10 38ns
recv 10 1.675µs
-----
recv 48 56.253µs
recv 58 48ns
recv 49 38ns
recv 10 38ns
recv 10 42.395253ms
-----
recv 48 176.57µs
recv 58 90ns
recv 49 59ns
recv 10 53ns
recv 10 43.771028ms
-----
recv 48 173.517µs
recv 58 89ns
recv 49 58ns
recv 10 56ns
recv 10 43.780377ms
-----
recv 48 160.039µs
recv 58 92ns
recv 49 57ns
recv 10 59ns
recv 10 43.75396ms
```

Stracing shows that it really is waiting in the epoll_pwait syscall for the daemon to send data.
```
socket(AF_INET, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = 3 <0.000017>
setsockopt(3, SOL_SOCKET, SO_BROADCAST, [1], 4) = 0 <0.000009>
connect(3, {sa_family=AF_INET, sin_port=htons(23456), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress) <0.000066>
epoll_create1(EPOLL_CLOEXEC)            = 4 <0.000009>
epoll_ctl(4, EPOLL_CTL_ADD, 3, {EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, {u32=1011842824, u64=139999765823240}}) = 0 <0.000010>
epoll_pwait(4, [{EPOLLOUT, {u32=1011842824, u64=139999765823240}}], 128, 0, NULL, 1) = 1 <0.000008>
getsockopt(3, SOL_SOCKET, SO_ERROR, [0], [4]) = 0 <0.000008>
getpeername(3, {sa_family=AF_INET, sin_port=htons(23456), sin_addr=inet_addr("127.0.0.1")}, [112->16]) = 0 <0.000008>
getsockname(3, {sa_family=AF_INET, sin_port=htons(34116), sin_addr=inet_addr("127.0.0.1")}, [112->16]) = 0 <0.000006>
setsockopt(3, SOL_TCP, TCP_NODELAY, [1], 4) = 0 <0.000007>
setsockopt(3, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0 <0.000007>
setsockopt(3, SOL_TCP, TCP_KEEPINTVL, [15], 4) = 0 <0.000008>
setsockopt(3, SOL_TCP, TCP_KEEPIDLE, [15], 4) = 0 <0.000007>
write(3, "shared |Context c:1.0\n|Action a:"..., 37) = 37 <0.000028>
read(3, 0xc0000bc000, 4096)             = -1 EAGAIN (Resource temporarily unavailable) <0.000006>
epoll_pwait(4, [], 128, 0, NULL, 0)     = 0 <0.000007>
epoll_pwait(4, [{EPOLLIN|EPOLLOUT, {u32=1011842824, u64=139999765823240}}], 128, -1, NULL, 0) = 1 <0.000374>
futex(0x5ff0d0, FUTEX_WAKE_PRIVATE, 1)  = 1 <0.000007>
futex(0x5fefd0, FUTEX_WAKE_PRIVATE, 1)  = 1 <0.000007>
read(3, "0:1\n\n", 4096)                = 5 <0.000012>
write(3, "shared |Context c:1.0\n|Action a:"..., 37) = 37 <0.000020>
read(3, 0xc0000bc000, 4096)             = -1 EAGAIN (Resource temporarily unavailable) <0.000007>
epoll_pwait(4, [{EPOLLOUT, {u32=1011842824, u64=139999765823240}}], 128, 0, NULL, 0) = 1 <0.000007>
epoll_pwait(4, [{EPOLLIN|EPOLLOUT, {u32=1011842824, u64=139999765823240}}], 128, -1, NULL, 0) = 1 <0.000007>
read(3, "0:1\n", 4096)                  = 4 <0.000009>
read(3, 0xc0000bc000, 4096)             = -1 EAGAIN (Resource temporarily unavailable) <0.000007>
epoll_pwait(4, [], 128, 0, NULL, 0)     = 0 <0.000007>
epoll_pwait(4, [{EPOLLIN|EPOLLOUT, {u32=1011842824, u64=139999765823240}}], 128, -1, NULL, 0) = 1 <0.040169>  <<<<<<<<< HERE <<<<<<<<<
futex(0x5ff0d0, FUTEX_WAKE_PRIVATE, 1)  = 1 <0.000023>
futex(0x5fefd0, FUTEX_WAKE_PRIVATE, 1)  = 1 <0.000013>
read(3, "\n", 4096)                     = 1 <0.000085>
write(3, "shared |Context c:1.0\n|Action a:"..., 37) = 37 <0.000072>
read(3, 0xc0000bc000, 4096)             = -1 EAGAIN (Resource temporarily unavailable) <0.000054>
epoll_pwait(4, [{EPOLLIN|EPOLLOUT, {u32=1011842824, u64=139999765823240}}], 128, 0, NULL, 0) = 1 <0.000010>
read(3, "0:1\n", 4096)                  = 4 <0.000012>
read(3, 0xc0000bc000, 4096)             = -1 EAGAIN (Resource temporarily unavailable) <0.000008>
epoll_pwait(4, [], 128, 0, NULL, 0)     = 0 <0.000008>
epoll_pwait(4, [{EPOLLIN|EPOLLOUT, {u32=1011842824, u64=139999765823240}}], 128, -1, NULL, 0) = 1 <0.047021>  <<<<<<<<< HERE <<<<<<<<<
```

#### Environment
```
$ vw --version
8.7.0 (git commit: 92aa2d71)

$ uname -isrv
Linux 5.0.0-36-generic #39-Ubuntu SMP Tue Nov 12 09:46:06 UTC 2019 x86_64
```

**Comment by jackgerrits** [comment]
Can you provide the Python snippet for reproing this?

**Comment by FilipVozar** [comment]
Sure, here it is:

```
import socket
import select
import time

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect(('127.0.0.1', 23456))
s.setblocking(0)

epoll = select.epoll()
epoll.register(s.fileno(), select.EPOLLOUT)

responses = 0

response = b''

while responses < 5:
    t = time.time()
    events = epoll.poll(1)
    for fileno, event in events:
        if event & select.EPOLLOUT:
            s.send('shared |Context f:1.0\n|Action a:1.0\n\n'.encode())
            epoll.modify(s.fileno(), select.EPOLLIN)
        elif event & select.EPOLLIN:
            resp = s.recv(1)
            print("read", resp, '{0:.4f}ms'.format((time.time() - t)*1000))
            response += resp
            if b'\n\n' in response:
                responses += 1
                epoll.modify(s.fileno(), select.EPOLLOUT)
                response = b''
        
epoll.unregister(s.fileno())
epoll.close()
s.close()
```

my output (python 3.7.3, but also works with 2.7.16):
```
read b'0' 0.5136ms
read b':' 0.0050ms
read b'1' 0.0031ms
read b'\n' 0.0029ms
read b'\n' 0.0036ms
read b'0' 0.1488ms
read b':' 0.0048ms
read b'1' 0.0031ms
read b'\n' 0.0033ms
read b'\n' 40.5414ms
read b'0' 0.2007ms
read b':' 0.0079ms
read b'1' 0.0052ms
read b'\n' 0.0055ms
read b'\n' 43.7713ms
read b'0' 0.1950ms
read b':' 0.0083ms
read b'1' 0.0050ms
read b'\n' 0.0052ms
read b'\n' 43.7155ms
read b'0' 0.1986ms
read b':' 0.0086ms
read b'1' 0.0105ms
read b'\n' 0.0052ms
read b'\n' 43.7098ms
```

**Comment by jackgerrits** [comment]
Great, thanks! I'll look into this

**Comment by FilipVozar** [comment]
I've just tried the same thing on my macbook, and noticed a different behaviour - the newlines are sent immediately, which is the behaviour I would expect. Most of the time I can read the whole response (including both trailing newlines). Sometimes the trailing newline is read in the next read call, but it's available immediately - unlike on linux, which has the ~40ms delay consistently.

I've modified the snippet to run on both linux and mac (since mac doesn't have epoll):

```
import socket
import select
import time

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect(('127.0.0.1', 23456))
s.setblocking(0)

responses = 0
response = b''

reading = []
writing = [s.fileno()]

while responses < 10:
    t = time.time()
    rd, wr, _ = select.select(reading, writing, [])
    if wr:
        s.send('shared |Context f:1.0\n|Action a:1.0\n\n'.encode())
        reading = [s.fileno()]
        writing = []
    if rd:
        resp = s.recv(1024)
        print("read", resp, '{0:.4f}ms'.format((time.time() - t)*1000))
        response += resp
        if b'\n\n' in response:
            responses += 1
            reading = []
            writing = [s.fileno()]
            response = b''
```

mac output:
```
read b'0:1\n\n' 0.1552ms
read b'0:1\n\n' 0.2110ms
read b'0:1\n\n' 0.1020ms
read b'0:1\n\n' 0.1111ms
read b'0:1\n\n' 0.1280ms
read b'0:1\n\n' 0.1049ms
read b'0:1\n' 0.0978ms
read b'\n' 0.0131ms
read b'0:1\n\n' 0.1240ms
read b'0:1\n\n' 0.0980ms
read b'0:1\n\n' 0.1328ms
```

linux output
```
read b'0:1\n\n' 0.2193ms
read b'0:1\n' 0.1493ms
read b'\n' 41.8534ms
read b'0:1\n' 0.2117ms
read b'\n' 43.7293ms
read b'0:1\n' 0.2098ms
read b'\n' 43.7660ms
read b'0:1\n' 0.1891ms
read b'\n' 43.7088ms
read b'0:1\n' 0.2158ms
read b'\n' 43.7372ms
read b'0:1\n' 0.2022ms
read b'\n' 43.7541ms
read b'0:1\n' 0.2079ms
read b'\n' 43.7431ms
read b'0:1\n' 0.2134ms
read b'\n' 43.7307ms
read b'0:1\n' 0.2103ms
read b'\n' 43.7481ms
```

mac vw version: `8.7.0` (installed simply by `brew install vowpal-wabbit`), OS version `10.15.1`

**Comment by jackgerrits** [comment]
I'm not able to repro this on my end using WSL Ubuntu 18.04.

```
read b'0:1\n' 2.5311ms
read b'\n' 0.0205ms
read b'0:1\n' 0.2713ms
read b'\n' 0.0172ms
read b'0:1\n' 0.2499ms
read b'\n' 0.0174ms
read b'0:1\n' 0.2136ms
read b'\n' 0.0162ms
read b'0:1\n' 0.2053ms
read b'\n' 0.0167ms
read b'0:1\n' 0.2177ms
read b'\n' 0.0165ms
read b'0:1\n' 0.2391ms
read b'\n' 0.0165ms
read b'0:1\n' 0.2110ms
read b'\n' 0.0167ms
read b'0:1\n' 0.2215ms
read b'\n' 0.0172ms
read b'0:1\n' 0.2072ms
read b'\n' 0.0160ms
```

The second newline is definitely sent in a separate write call. So, to hopefully reduce impact of this I'll see about getting the writes bundled together.

**Comment by FilipVozar** [comment]
I've built the binary statically on my machine
```
cmake .. -DSTATIC_LINK_VW=ON -DCMAKE_BUILD_TYPE=Release
make vw-bin -j `nproc`
```

put it into a Docker image, and ran it on Kubernetes cluster, just to see if it behaves differently on some other machine. I've got the same kind of delay (40+ms).
Unfortunately I don't know enough about network programming to debug this further - if you have any ideas for more tests and what to look for, please tell me and I'll try to test it more.

I'm hopeful that bundling the second newline together in a single write call could reduce the impact.

**Comment by TheTrope** [comment]
I'm dealing with real time low latency prediction, and I had the same issue a few moment ago.

What I discovered is that Vowpal Wabbit always struggle to send the last line. I had those 40ms delays for the last line of the chunk I sent. (Not sure this has a 100% reproduction rate)
However, if I send a request before this delays end, Vowpal Wabbit somehow wake up and send the previous remaining one. 
The workaround I've found so far is to spam VW every 1ms with fake empty requests in order to force it to answer. I guess the TCP server is somehow faulty (Or maybe it is some kind of optimisation)

Ubuntu 18
